code/iaas/auth-logic/src/main/java/io/cattle/platform/iaas/api/auth/dao/impl/AuthDaoImpl.java

```diff
@@ -129,70 +160,221 @@ public void updateAccount(Account account, String name, String kind, String exte
-                                            .and(ACCOUNT.STATE.ne("purged"))
-                                    .orderBy(ACCOUNT.ID.asc()).fetch();
-        
-        for(AccountRecord orgProject: orgProjects) {
-            projects.add(orgProject);
-        }
-        
-        for(AccountRecord teamProject: teamProjects) {
-            projects.add(teamProject);
-        }
-        
-        projects.addAll(create()
-                .selectFrom(ACCOUNT)
-                .where(ACCOUNT.EXTERNAL_ID.eq(userId)
-                        .and(ACCOUNT.EXTERNAL_ID_TYPE.eq("project:github_user")))
-                        .and(ACCOUNT.STATE.ne("purged"))
-                        .fetch());
-        
+    public List<Account> getAccessibleProjects (Set<ExternalId> externalIds, Boolean isAdmin, Long policyAccount) {
+        List<Account> projects = new ArrayList<>();
+        if(externalIds == null) {
+            return projects;
+        }
+        if (isAdmin) {
+            projects.addAll(create()
+                    .selectFrom(ACCOUNT)
+                    .where(ACCOUNT.KIND.eq(ProjectConstants.TYPE)
+                    .and(ACCOUNT.STATE.eq(CommonStatesConstants.ACTIVE)))
+                    .orderBy(ACCOUNT.ID.asc())
+                    .fetch());
+            return projects;
+        }
+        if (policyAccount != null) {
+            Account account = getAccountById(policyAccount);
+            if (account != null) {
+                if (account.getKind().equalsIgnoreCase(ProjectConstants.TYPE)) {
+                    projects.add(account);
+                }
+            }
+        }
+        //DSL.falseCondition is created so that we can dynamically build a or
+        //Condition without caring what the external Ids are and still make one
+        //Database call.
+        Condition allMembers = DSL.falseCondition();
+        for (ExternalId id: externalIds){
+            allMembers = allMembers.or(PROJECT_MEMBER.EXTERNAL_ID.eq(id.getId())
+                    .and(PROJECT_MEMBER.EXTERNAL_ID_TYPE.eq(id.getType()))
+                    .and(PROJECT_MEMBER.REMOVED.isNull())
+                    .and(PROJECT_MEMBER.STATE.eq(CommonStatesConstants.ACTIVE)));
+        }
+        SelectQuery query = create().selectQuery();
+        query.addFrom(ACCOUNT);
+        query.addJoin(PROJECT_MEMBER, PROJECT_MEMBER.PROJECT_ID.equal(ACCOUNT.ID));
+        query.addConditions(allMembers);
+        query.setDistinct(true);
+        projects.addAll(query.fetchInto(ACCOUNT));
         return projects;
-        
+    }
+
+    public List<? extends ProjectMember> getProjectMembers(Long id) {
+        return create()
+                .selectFrom(PROJECT_MEMBER)
+                .where(PROJECT_MEMBER.PROJECT_ID.eq(id))
+                .and(PROJECT_MEMBER.STATE.eq(CommonStatesConstants.ACTIVE))
+                .and(PROJECT_MEMBER.REMOVED.isNull())
+                .orderBy(PROJECT_MEMBER.ID.asc()).fetch();
+    }
+
+    @Override
```

Comment:
You do implicitly. Take this code as an example (ProjectMemberResourceManager.java):

ProjectMember projectMember = authDao.getProjectMember(Long.valueOf(id));
Then further down in Dao calls invoked by hasAccessToProject, you always check that the project member is in active state:

+            if (!authDao.hasAccessToProject(projectMember.getProjectId(), policy.getAccountId(),
+                    policy.isOption(Policy.AUTHORIZED_FOR_ALL_ACCOUNTS), policy.getExternalIds())) {
+                throw new ClientVisibleException(ResponseCodes.NOT_FOUND);
+            }


Eventually it all comes to isProjectMember dao call where you make the check:

 .and(PROJECT_MEMBER.STATE.eq(CommonStatesConstants.ACTIVE))

As only active project member Checking it before hand by requesting acrive project member, will spare you from making all subsequent dao db calls.